CosmicReco/inc/MinuitDriftFitter.hh

```diff
@@ -52,7 +52,7 @@ struct FitResult{
 
 namespace MinuitDriftFitter {
 	
-	FitResult DoFit(int diag, CosmicTrackFinderData& trackdata, StrawResponse::cptr_t srep, const std::vector<Straw const& > straws, const Tracker* tracker, double doca_cut, unsigned int MinNCh_cut, int LogLcut, double _gaussTres, double _maxTres);
+	FitResult DoFit(int diag, CosmicTrackFinderData& trackdata, StrawResponse::cptr_t srep, const std::vector<Straw> straws, const Tracker* tracker, double doca_cut, unsigned int MinNCh_cut, int LogLcut, double _gaussTres, double _maxTres);
```

Comment:
The third argument should be of reference type:

std::vector<Straw> const& straws

As it is now you are making a copy of the entire straw geometry when you call the function.

Reply:
Hi, 

> Thanks Ryun, that's correct. Vector semantics require default constructor, assignment etc for its template argument, those can't be supplied for type &. Passing a vector of pointers works but those are ugly to work with. The alternative design for passing a subset of straws is to pass a vector of StrawIDs and the Tracker object itself, then call getStraw(StrawID) at the point of use. That's more consistent with the design, but has a cost of the additional lookup. I believe that cost is negligible as the lookup uses a hash. Dave
> […](#)
> On Fri, Jan 10, 2020 at 6:40 AM ryuwd ***@***.***> wrote: Hi Sophie, I came across that problem a while ago, if you want to pass a collection of Straws +avoid deep copying you could try std::reference_wrapper or using pointers (const Straw *) Best wishes, Ryun On Fri, 10 Jan 2020 at 13:47, Sophie Middleton ***@***.***> wrote: > ***@***.**** commented on this pull request. > ------------------------------ > > In CosmicReco/src/DriftFitUtils.cc > <[#94 (comment)](https://github.com/Mu2e/Offline/pull/94#discussion_r365240828)>: > > > + > + TwoLinePCA_XYZ PCA = TwoLinePCA_XYZ(track_position, > + track_direction, > + wire_position, > + wire_direction, > + 1.e-8); > + > + double ambig = PCA.LRambig(); > + > + return ambig; > + > +} > + > + > +int DriftFitUtils::GetAmbig(ComboHit chit, double a0, double a1, double b0, double b1) { > + mu2e::GeomHandle<mu2e::Tracker> th; > > Sorted it now > > — > You are receiving this because you are subscribed to this thread. > Reply to this email directly, view it on GitHub > < #94?email_source=notifications&email_token=ANOMGYW6OWXTM4QNSGHBWUTQ5B4ATA5CNFSM4KE2RYSKYY3PNVWWK3TUL52HS4DFWFIHK3DMKJSXC5LFON2FEZLWNFSXPKTDN5WW2ZLOORPWSZGOCRK5Z3Y#discussion_r365240828 >, > or unsubscribe > < https://github.com/notifications/unsubscribe-auth/ANOMGYQPINPEJHLHG4W7D6DQ5B4ATANCNFSM4KE2RYSA > > . > — You are receiving this because you were assigned. Reply to this email directly, view it on GitHub <#94?email_source=notifications&email_token=ABAH57Z7LMYTLSQYHODY5BDQ5CCERA5CNFSM4KE2RYSKYY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOEIUDP4A#issuecomment-573061104>, or unsubscribe <https://github.com/notifications/unsubscribe-auth/ABAH574U7OYUQ765NMSWJB3Q5CCERANCNFSM4KE2RYSA> .
> -- David Nathan Brown Dave_Brown@lbl.gov Office Phone (510) 486-7261 Fax 495-2957 Lawrence Berkeley National Lab MS 50R5008 (50-6026C) Berkeley, CA 94720

So should I do as Dave says and have a vector of strawIDs and the tracker object? I think it would be good to get this in its best form now as I've played around with various versions. What is the best recommendation for speed/best coding practice?

Reply:
Oh, maybe I missunderstood - should I just use std::vector<Straw> const& straws ?